### PR TITLE
Update generate-all-docs command logic to fetch installed plugins

### LINKED ISSUE #1765: Non-context aware plugin listing is broken
**Bug description**

There are a few spots in the code that have not been updated for the context aware plugin handling. One of those is the `tanzu generate-all-docs` command. Running this command removes all references to subcommands from the generated CLI docs because it doesn't find any plugins.

The places in the code where this had previously been called that are still working have been updated to have something like:

```go
if config.IsFeatureActivated(config.FeatureContextAwareCLIForPlugins) {
		currentServerName := ""

		server, err := config.GetCurrentServer()
		if err == nil && server != nil {
			currentServerName = server.Name
		}

		serverPlugin, standalonePlugins, err := pluginmanager.InstalledPlugins(currentServerName)
		if err != nil {
			return nil, fmt.Errorf("find installed plugins: %w", err)
		}

		//nolint:gocritic
		p := append(serverPlugin, standalonePlugins...)
		for i := range p {
			plugins = append(plugins, &p[i])
		}
	} else {
		plugins, err = cli.ListPlugins()
		if err != nil {
			return nil, fmt.Errorf("find available plugins: %w", err)
		}
	}
```

But other places, like the generate-all-docs code, just call `ListPlugins` to get the set of installed plugins. That fails to find any because it looks like the plugin cache has been changed from being `~/.local/share/tanzu-cli/[set of binaries]` to be `~/.local/share/tanzu-cli/[plugin_name]/[version_named_binary]`.

The `ListPlugins` code explicitly checks if the first level under `~/.local/share/tanzu-cli/` is a directory, and if so ignores it and continues:

https://github.com/vmware-tanzu/tanzu-framework/blob/3ad462fa5fa60a336153442fbda17047d53d1ac1/pkg/v1/cli/catalog.go#L244

Since it appears this can now contain multiple versions per plugin directory, special handling would need to be added to look for the most recent binary version.

Some of the places where there is no conditional handling for context aware plugins:

https://github.com/vmware-tanzu/tanzu-framework/blob/7120a043b0c738f376810591037430ed6a0c8949/cmd/cli/plugin-admin/test/main.go#L39
https://github.com/vmware-tanzu/tanzu-framework/blob/7120a043b0c738f376810591037430ed6a0c8949/pkg/v1/cli/command/core/doc.go#L42
https://github.com/vmware-tanzu/tanzu-framework/blob/7120a043b0c738f376810591037430ed6a0c8949/pkg/v1/cli/catalog.go#L552
https://github.com/vmware-tanzu/tanzu-framework/blob/7120a043b0c738f376810591037430ed6a0c8949/pkg/v1/cli/command/core/config.go#L235
https://github.com/vmware-tanzu/tanzu-framework/blob/7120a043b0c738f376810591037430ed6a0c8949/pkg/v1/cli/command/core/plugin_manager.go#L106
https://github.com/vmware-tanzu/tanzu-framework/blob/7120a043b0c738f376810591037430ed6a0c8949/pkg/v1/cli/command/core/update.go#L44

**Affected product area (please put an X in all that apply)**

- [ ] APIs
- [ ] Addons
- [X] CLI
- [X] Docs
- [ ] IAM
- [ ] Installation
- [X] Plugin
- [ ] Security
- [ ] Test and Release
- [ ] User Experience

**Expected behavior**

Non-context aware plugins get returned for any instance of `ListPlugins` and that `tanzu generate-all-docs` generates the docs for all CLI commands.

### PR DESCRIPTION
**Pull request by chandrareddyp**
### What this PR does / why we need it
The `tanzu generate-all-docs` command is broken because it's using the legacy API to fetch installed plugins and does not consider user input value for the output directory; this PR fixes below:
1. `tanzu generate-all-docs` command logic updated to use context-aware plugin list API, and deprecate the legacy API
2. `tanzu generate-all-docs --docs-dir` command logic updated to consider the user input value for the input flag '--docs-dir' docs output directory and generate docs to the given output directory
3. update the error message if there is any error while reading the output directory

### Which issue(s) this PR fixes

Fixes #1765

### Describe testing done for PR

- Manual testing is done for `tanzu generate-all-docs` to test docs for all installed plugins
- Manual testing is done for `tanzu generate-all-docs --docs-dir` to test docs to be generated on the given output directory
- Unit test cases are updated

### Release note

```release-note
`tanzu generate-all-docs` command implementation has been updated to use context-aware plugin list API and generate docs to given output folder
```

### CODE REVIEW

File: pkg/v1/cli/command/core/doc.go
Comment (anujc25) [inline pkg/v1/cli/command/core/doc.go:42]: We should continue to use this conditionally `config.IsFeatureActivated(config.FeatureContextAwareCLIForPlugins)` similar to what we are doing in other places until we completely remove deprecated code?

Comment (chandrareddyp) [reply pkg/v1/cli/command/core/doc.go:42]: changes are done to use conditional flag.

Comment (anujc25) [reply pkg/v1/cli/command/core/doc.go:42]: @chandrareddyp what I mean here is to continue using `ListPlugins` when featureflag is false and not remove it completely to align the code with other conditional changes.

```
if !config.IsFeatureActivated(config.FeatureContextAwareCLIForPlugins){
    pluginDescriptions, err := cli.ListPlugins()
} else {
    pluginDescriptions, err = pluginmanager.InstalledPluginsDescriptors() 
}
```

Comment (chandrareddyp) [reply pkg/v1/cli/command/core/doc.go:42]: got it. will update the code. thanks.

File: pkg/v1/cli/command/core/doc.go
Comment (marckhouzam) [inline pkg/v1/cli/command/core/doc.go:47]: I suggest we also check that `docsDir` is a directory.
Currently:
```
$ touch tmpDoc. # Creates a file, not a directory
$ tanzu generate-all-docs --docs-dir tmpDoc
✖  error generate core tanzu cmd markdown "error opening tanzu.md \"open tmpDoc/tanzu.md: not a directory\""
```

File: pkg/v1/cli/command/core/doc.go
Comment (marckhouzam) [inline pkg/v1/cli/command/core/doc.go:58]: Should we continue using the deprecated function, or should we use the recommended replacement?

File: pkg/v1/cli/command/core/doc.go
Comment (marckhouzam) [inline pkg/v1/cli/command/core/doc.go:106]: Nice!

**Review by marckhouzam — COMMENTED** [review]
Very nice.  LGTM.

Just a couple of nits and suggestions.

**Review by mpanchajanya — APPROVED** [review]

File: pkg/v1/cli/command/core/doc.go
Comment (chandrareddyp) [reply pkg/v1/cli/command/core/doc.go:47]: sure, i will update to check input path is a directory or not.

File: pkg/v1/cli/command/core/doc.go
Comment (chandrareddyp) [reply pkg/v1/cli/command/core/doc.go:26]: thanks, i will update.

File: pkg/v1/cli/command/core/doc.go
Comment (chandrareddyp) [reply pkg/v1/cli/command/core/doc.go:58]: there are some git integration tests still depend on deprecated code, I have created new issue to address it and remove the deprecated code - https://github.com/vmware-tanzu/tanzu-framework/issues/3224

Comment (marckhouzam) [reply pkg/v1/cli/command/core/doc.go:58]: > there are some git integration tests still depend on deprecated code, I have created new issue to address it and remove the deprecated code - #3224

Sorry, I did't express myself well.  We can keep the deprecated function `ListPlugins`, but in line 58 above, do we still need to use it or can we use the replacement function?

File: pkg/v1/cli/command/plugin/doc.go
Comment (marckhouzam) [inline pkg/v1/cli/command/plugin/doc.go:35]: Here too I think we should check that `docsDir` is a directory

**Review by marckhouzam — COMMENTED** [review]
Thanks for the changes.

This PR will have a minor conflict with @anujc25 large refactoring of #3201.  I think adapting this PR will be easier than the other way around?  If so, maybe we want to wait until #3201 is merged?

File: pkg/v1/cli/command/core/doc.go
Comment (chandrareddyp) [reply pkg/v1/cli/command/core/doc.go:58]: We still need to use it, initially I have removed it, later Anuj suggested better to keep it until we remove the deprecated code completely.

File: pkg/v1/cli/command/plugin/doc.go
Comment (chandrareddyp) [reply pkg/v1/cli/command/plugin/doc.go]: yes thats correct.

**Review by marckhouzam — APPROVED** [review]
LGTM

**Review by anujc25 — APPROVED** [review]
LGTM. Thanks for the fix.

File: cli/runtime/plugin/doc.go
Comment (anujc25) [inline cli/runtime/plugin/doc.go:47]: Looks like incorrect merge conflict resolution as it adds `core` as a dependency.

File: cli/runtime/plugin/doc.go
Comment (anujc25) [inline cli/runtime/plugin/doc.go:16]: `cli/runtime` module should not have dependency on root level tanzu-framework module

**Pull request merged**
